src/Rector/ClassMethod/AddGenericReturnTypeToRelationsRector.php

```diff
@@ -0,0 +1,177 @@
+    {
+        return new RuleDefinition(
+            'Add generic return type to relations in child of Illuminate\Database\Eloquent\Model',
+            [
+                new CodeSample(
+                    <<<'CODE_SAMPLE'
+use App\Account;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Relations\HasMany;
+
+class User extends Model
+{
+    public function accounts(): HasMany
+    {
+        return $this->hasMany(Account::class);
+    }
+}
+CODE_SAMPLE
+
+                    ,
+                    <<<'CODE_SAMPLE'
+use App\Account;
+use Illuminate\Database\Eloquent\Model;
+use Illuminate\Database\Eloquent\Relations\HasMany;
+
+class User extends Model
+{
+    /** @return HasMany<Account> */
+    public function accounts(): HasMany
+    {
+        return $this->hasMany(Account::class);
+    }
+}
+CODE_SAMPLE
+                ),
+            ]
+        );
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [ClassMethod::class];
+    }
+
+    /**
+     * @param ClassMethod $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if ($node->stmts === null) {
+            return null;
+        }
+
+        $classLike = $this->betterNodeFinder->findParentType($node, ClassLike::class);
+
+        if (! $classLike instanceof ClassLike) {
+            return null;
+        }
+
+        if (! $this->isObjectType($classLike, new ObjectType('Illuminate\Database\Eloquent\Model'))) {
+            return null;
+        }
+
+        $methodReturnType = $node->getReturnType();
+
+        if ($methodReturnType === null) {
+            return null;
+        }
+
+        $methodReturnTypeName = $this->getName($methodReturnType);
+
+        if ($methodReturnTypeName === null) {
+            return null;
+        }
+
+        if (! $this->isObjectType(
+            $methodReturnType,
+            new ObjectType('Illuminate\Database\Eloquent\Relations\Relation')
+        )) {
+            return null;
+        }
+
+        $docInfo = $this->phpDocInfoFactory->createFromNode($node);
+        if (! $docInfo instanceof PhpDocInfo) {
+            return null;
+        }
+
+        // Return, if already has return type
+        if ($node->getDocComment() !== null && $docInfo->hasByName('return')) {
+            return null;
+        }
+
+        $returnStatement = $this->betterNodeFinder->findFirstInstanceOf($node->stmts, Return_::class);
+
+        if (! $returnStatement instanceof Return_) {
+            return null;
+        }
+
+        $relationMethodCall = $this->betterNodeFinder->findFirstInstanceOf($returnStatement, MethodCall::class);
+
+        if (! $relationMethodCall instanceof MethodCall) {
+            return null;
+        }
+
+        $relatedClass = $this->getRelatedModelClassFromMethodCall($relationMethodCall);
+
+        if ($relatedClass === null) {
+            return null;
+        }
```

Comment:
please move some return null code block above to separate private methods based on the code purpose to reduce complexity, eg:

```
private function shouldSkipByObjectType() ...
private function shouldSkipByReturnType() ...
```

Reply:
Thank you for the tip! It's reduced now.